rfcs/0043-backward-compatibility-of-stream-plan.md

```diff
@@ -0,0 +1,39 @@
+- in other databases, the plan node is just used as a communicating protocol between frontend and compute node. So the compute node can only support the latest version's plan node format and reject all the requests with unknown plan node. But our stream plan should be persistent in meta store which means that a compute node must be compatible with all versions of old plans' protobuf format.
+
+In conclusion, we need find a way to achieve a balance between rapid development and backward compatibility, especially for stream plan node.
+
+## Design
+
+### Nightly and Stable SQL Features 
+Distinguish the nightly and stable feature when publishing release version. RW will do not ensure compatibility for the streaming jobs with the nightly features in following releases. For example, if we release the "emit on close" as a in the release v0.1.17 and user create a mv with that feature on a v0.1.17 cluster. The v0.1.18 and following version's RW can not ensure it can run successfully on the existing streaming jobs. User can drop the MVs with the nightly feature before they upgrade the cluster. For those nightly feature users really what to upgrade, we can write helper scripts too. And the stable features will be tested with new released compute node on old version streaming plans. Also, with the convinced stable feature list, we can test the backward compatibility more easily. 
+
+### Nightly and Stable Stream Plan Node
+How to know if a SQL Feature has been stable? Developer should comment the compatibility annotation on protobuf struct of every stream plan node(like annotation in java). the annotation contains: "nightly v0.1.14", "stable v0.1.15", "deprecated v0.1.16". A plan node will be with a nightly annotation firstly. When developer ensure that the plan node struct is stable enough, a stable annotation should be comments on the protobuf struct. When developer ensure that frontend will not generate the plan node, a deprecated annotation should be comments on the protobuf struct. A SQL feature is stable means that all the stream plan nodes generated by any version's optimizer should have been stable.
+
+To be discussed: what is the proper format of those comments in proto files and how to check all plan node should have one in CI check?
+
+### Copy-on-Write Style Changes on Stable Plan Node Protobuf
+How to maintain the compatibility of the plan node's protobuf? If developer want to do any changes on a stable plan node, he should add a new plan node protobuf definition. For example, if he want to add a new field in `StreamHashAgg`, he must define a new protobuf struct `StreamHashAggV2` and add the field on that. Notice that there are multi versions protobuf but they can share the same implementation.  
```

Comment:
Should we keep all `StreamHashAgg` versions even after the release?
If we have two versions of `StreamHashAggV2` in a release, will CN also handle V1?

Reply:
Yes, because the any old meta store can store a stream plan with `StreamHashAggV1`

Reply:
hmm in this case, will add a version field in `StreamHashAgg` a valid choice?

Reply:
it is a special case that we just add a new field and I am not sure if it is general enough